pkg/client/introduce/client.go

```diff
@@ -49,80 +57,141 @@ func New(ctx Provider) (*Client, error) {
 	return c.handleOutbound(&introduce.Request{
 		Type: introduce.RequestMsgType,
-		ID:   uuid.New().String(),
+		ID:   c.newUUID(),
+	}, InvitationEnvelope{
+		Dests: []*service.Destination{dest},
 	})
 }
 
-// HandleRequest executes after receiving a request (the client does not have a public Invitation)
-func (c *Client) HandleRequest(dest1, dest2 *service.Destination) error {
-	return c.sendProposal(options{
-		Destinations: []*service.Destination{dest1, dest2},
+// HandleRequest is a helper function to prepare the right protocol dependency interface
+// It can be executed after receiving a Request action message (the client does not have a public Invitation)
+func (c *Client) HandleRequest(msg service.DIDCommMsg, dest1, dest2 *service.Destination) error {
+	thID, err := msg.ThreadID()
+	if err != nil {
+		return fmt.Errorf("handle request threadID: %w", err)
+	}
+
+	return c.saveInvitationEnvelope(thID, InvitationEnvelope{
+		Dests: []*service.Destination{dest1, dest2},
 	})
 }
 
-// HandleRequestWithInvitation executes after receiving a request when the client has a public Invitation
-func (c *Client) HandleRequestWithInvitation(inv *didexchange.Invitation, dest *service.Destination) error {
-	return c.sendProposal(options{
-		Invitation:   inv,
-		Destinations: []*service.Destination{dest},
+// HandleRequestWithInvitation is a helper function to prepare the right protocol dependency interface
+// It can be executed after receiving a Request action message (the client has a public Invitation)
+// nolint: lll
+func (c *Client) HandleRequestWithInvitation(msg service.DIDCommMsg, inv *didexchange.Invitation, dest *service.Destination) error {
+	thID, err := msg.ThreadID()
+	if err != nil {
+		return fmt.Errorf("handle request with invitation threadID: %w", err)
+	}
+
+	return c.saveInvitationEnvelope(thID, InvitationEnvelope{
+		Inv:   inv,
+		Dests: []*service.Destination{dest},
 	})
 }
 
-func (c *Client) handleOutbound(msg interface{}) error {
+func (c *Client) handleOutbound(msg interface{}, o InvitationEnvelope) error {
 	payload, err := json.Marshal(msg)
 	if err != nil {
-		return err
+		return fmt.Errorf("marshal outbound msg: %w", err)
 	}
 
 	didMsg, err := service.NewDIDCommMsg(payload)
 	if err != nil {
+		return fmt.Errorf("new outbound DIDCommMsg msg: %w", err)
+	}
+
+	thID, err := didMsg.ThreadID()
+	if err != nil {
+		return fmt.Errorf("outbound threadID: %w", err)
+	}
+
+	if err := c.saveInvitationEnvelope(thID, o); err != nil {
 		return err
 	}
 
-	return c.service.HandleOutbound(didMsg, &service.Destination{})
+	return c.service.HandleOutbound(didMsg, o.Dests[0])
 }
 
-// options keeps the information needed for sending a proposal
-type options struct {
+// InvitationEnvelope keeps the information needed for sending a proposal
+type InvitationEnvelope struct {
 	// Invitation must be set when we have a public Invitation
-	Invitation *didexchange.Invitation
+	Inv *didexchange.Invitation `json:"inv,omitempty"`
```

Comment:
Do we need to expose these fields ?

Reply:
Yeah, for marshall, this structure is being stored in DB